core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -650,6 +653,28 @@ else if (redistributeWrites) {
                     partitioningScheme = Optional.of(new PartitioningScheme(Partitioning.create(FIXED_ARBITRARY_DISTRIBUTION, ImmutableList.of()), newSource.getNode().getOutputSymbols()));
                 }
             }
+            else if (scaleWriters
+                    && writerTarget.supportsReportingWrittenBytes(plannerContext.getMetadata(), session)
+                    && writerTarget.supportsMultipleWritersPerPartition(plannerContext.getMetadata(), session)
+                    // do not insert an exchange if partitioning is compatible
+                    && !newSource.getProperties().isCompatibleTablePartitioningWith(partitioningScheme.get().getPartitioning(), false, plannerContext.getMetadata(), session)) {
+                if (partitioningScheme.get().getPartitioning().getHandle().equals(FIXED_HASH_DISTRIBUTION)) {
+                    partitioningScheme = Optional.of(partitioningScheme.get().withPartitioningHandle(SCALED_WRITER_HASH_DISTRIBUTION));
+                }
+                else {
+                    PartitioningHandle partitioningHandle = partitioningScheme.get().getPartitioning().getHandle();
+                    verify(!(partitioningHandle.getConnectorHandle() instanceof SystemPartitioningHandle));
+                    verify(
```

Comment:
why are we sure that this will never be the case?

Reply:
Basically what these lengthy checks are saying is that a `TableWriteNode` should either have no partitioning requirement, or if there's a requirement it must be a hash based partitioning. And hash based partitioning is either possible through a system handle (`FIXED_HASH_DISTRIBUTION`) or through a connector specific partitioning, that is assumed to be hash based.